diff --git a/lib/project_razor/broker/puppet/agent_install.erb b/lib/project_razor/broker/puppet/agent_install.erb
index 3f8af5a..573513f 100644
--- a/lib/project_razor/broker/puppet/agent_install.erb
+++ b/lib/project_razor/broker/puppet/agent_install.erb
@@ -2,7 +2,13 @@
 
 set -u
 set -e
-
+function install_eatmydata(){
+export DEBIAN_FRONTEND=noninteractive
+apt-get install -y eatmydata
+cat > /etc/profile.d/eatmydata.sh <<EOF
+export LD_PRELOAD=/usr/lib/libeatmydata/libeatmydata.so
+EOF
+}
 function install_puppet() {
     # TODO add support for versions
     gem install puppet --no-ri --no-rdoc
@@ -54,6 +60,8 @@ function facts_dot_d() {
 }
 
 function provision_puppet() {
+    install_eatmydata
+    export LD_PRELOAD=/usr/lib/libeatmydata/libeatmydata.so
     install_puppet
     configure_puppet
     <%= 'facts_dot_d' if @options[:metadata] %>
@@ -62,4 +70,5 @@ function provision_puppet() {
     exit 0
 }
 
-provision_puppet 2>&1 | tee /tmp/puppet.log
+provision_puppet
+
